### PR TITLE
Adds Subscription app in features, exposes API

### PR DESCRIPTION
**Pull request by rehan99000**
**Description:**
This PR adds a subscription app in open edx features and exposes APIs to get/create/update `UserSubscription` records.

**JIRA:**
https://edlyio.atlassian.net/browse/EDLY-1818

### CODE REVIEW

File: openedx/features/subscriptions/models.py
Comment (ziafazal) [inline openedx/features/subscriptions/models.py:49]: expiration date should either be null(full time access) or >= today.

File: openedx/features/subscriptions/models.py
Comment (ziafazal) [inline openedx/features/subscriptions/models.py:15]: This should be `subscription_uuid` instead of id.
and add `db_index=True`

File: openedx/features/subscriptions/api/v1/urls.py
Comment (ziafazal) [inline openedx/features/subscriptions/api/v1/urls.py:6]: nit: `name='subscriptions-list'`

File: openedx/features/subscriptions/api/v1/urls.py
Comment (ziafazal) [inline openedx/features/subscriptions/api/v1/urls.py:7]: nit: `name=subscriptions-retrieve-update`

File: openedx/features/subscriptions/admin.py
Comment (ziafazal) [inline openedx/features/subscriptions/admin.py:14]: add `search_fields = ['user__email', 'user__username']`

**Review by ziafazal — CHANGES_REQUESTED** [review]
@rehan99000 overall looks good have few minor suggestions.

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py:15]: @ziafazal A subscription does not have a uuid. It does however has a sku. If u want me to use that instead of this. @zubair-arbi mentioned this as well but as there was no uuid as it is a simple `Product`, he said this is fine for now.

File: openedx/features/subscriptions/api/v1/urls.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/urls.py:7]: `SUBSCRIPTION_URLS` has `'subscriptions'` passed as namespace so when i do `include(SUBSCRIPTION_URLS)`, it should provide a syntax like `subscriptions:list` and `subscriptions:retrieve_update`. If u want me to change that to this I can do that as well.

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py:49]: does that mean that expired `UserSubscriptions` would have null as `expiration_date` ?

File: openedx/features/subscriptions/admin.py
Comment (rehan99000) [reply openedx/features/subscriptions/admin.py:15]: done

File: openedx/features/subscriptions/models.py
Comment (ziafazal) [reply openedx/features/subscriptions/models.py:49]: no. I mean for those subscription which are not time bound in those we'd have null in `expiration_date`.

File: openedx/features/subscriptions/models.py
Comment (ziafazal) [inline openedx/features/subscriptions/models.py:16]: In which case `sku` is null?

Comment (rehan99000) [reply openedx/features/subscriptions/models.py:16]: It shouldn't be null in case of successful creation of a subscription product. However, the existing API for Course modes did it so I followed pattern.

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py:49]: Oh got it. let me update it

File: lms/envs/common.py
Comment (ziafazal) [inline lms/envs/common.py:2336]: We should make `subscription` a [plugin](https://github.com/edx/edx-platform/blob/master/docs/guides/extension_points.rst#platform-features-integrating-python-code) 
https://github.com/edx/edx-django-utils/blob/master/edx_django_utils/plugins/README.rst

File: lms/urls.py
Comment (ziafazal) [inline lms/urls.py:671]: We should make `subscription` a [plugin](https://github.com/edx/edx-platform/blob/master/docs/guides/extension_points.rst#platform-features-integrating-python-code) 
https://github.com/edx/edx-django-utils/blob/master/edx_django_utils/plugins/README.rst

File: openedx/features/subscriptions/models.py
Comment (ziafazal) [inline openedx/features/subscriptions/models.py:46]: we need to change the logic in this method get_active_subscription  . to determine if a user’s subscription is active or not
1. In case of (limited time and unlimited courses) subscription its expiry date should be greater than or equal to today.
2. In case of (unlimited time and limited courses) subscription it should have one or more courses enrollments available to be utilized
3. In case of (limited time and limited courses) its expiry date should be greater than or equal to today and it should have one or more course enrollments available to be utilized.
4. In case (unlimited time and unlimited courses) it is alway active.

Comment (rehan99000) [reply openedx/features/subscriptions/models.py]: @ziafazal Done.

File: openedx/features/subscriptions/api/v1/serializers.py
Comment (iam-mhaseeb) [inline openedx/features/subscriptions/api/v1/serializers.py:11]: `required=False` would filter out `course_enrollments` which means UserSubscription object will be created without `course_erollments`. Shouldn't we restrict it?

File: openedx/features/subscriptions/api/v1/views.py
Comment (iam-mhaseeb) [inline openedx/features/subscriptions/api/v1/views.py:32]:
```
return UserSubscription.objects.filter(user_id=user_id)
```
because you already did `user_id = self.request.query_params.get('user', self.request.user.id)`

Comment (iam-mhaseeb) [inline openedx/features/subscriptions/api/v1/views.py:32]: What is the case of non valid `UserSubscription`? Can you share some example?

File: openedx/features/subscriptions/api/v1/views.py
Comment (iam-mhaseeb) [inline openedx/features/subscriptions/api/v1/views.py:48]: Why first? A user can have multiple subscriptions.

File: openedx/features/subscriptions/utils.py
Comment (iam-mhaseeb) [inline openedx/features/subscriptions/utils.py:22]: What if the same course in multiple subscriptions? Would the user get all subscriptions?

**Review by iam-mhaseeb — CHANGES_REQUESTED** [review]
A few questions & 1 change requested.

File: openedx/features/subscriptions/api/v1/serializers.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/serializers.py:11]: @iam-mhaseeb This is for when a user buys the subscription. at that point, a user won't have any enrollments. This record is created through subscription fulfillment module which runs when you buy a new subscription. Subscription buy flow and enroll through subscription flows are different.

File: openedx/features/subscriptions/api/v1/views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py:32]: yes let me make this correct

Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py:32]: A subscription that is active but one can't use it to buy any new course seats. for example a subscription with course limit 5 and expires in 2 months. if a user enrolls in 5 courses within one month, the subscription is still active but is not valid meaning one can't use it to buy any new course. That's the major difference between active and valid subscription

File: openedx/features/subscriptions/api/v1/views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py:48]: This is a retrieve/update/create view. Also, kwargs have `lookup_field` as key which is `subscription_id`. There can only be one subscription against a subscription id for a user at max. but I need to filter it on user as well. going to make that change

File: openedx/features/subscriptions/utils.py
Comment (rehan99000) [reply openedx/features/subscriptions/utils.py:22]: I was a little unsure about this myself. We don't have the renew feature in this part but considering we would have it in future, I think it would be better if we check active for all such subscriptions. let me make this change as well

File: openedx/features/subscriptions/api/v1/views.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/api/v1/views.py:45]: I think we also need to add default order base of `created_at`
https://www.django-rest-framework.org/api-guide/filtering/#specifying-a-default-ordering

File: openedx/features/subscriptions/api/v1/views.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/api/v1/views.py:27]: `valid_subscriptions` -> `return_only_valid_subscriptions`

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
I have added my feedback.

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py]: changed this condition

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py:70]: done

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py]: done

File: openedx/features/subscriptions/utils.py
Comment (rehan99000) [reply openedx/features/subscriptions/utils.py:19]: done

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py]: yeah this condition didn't seem right. have removed it

File: openedx/features/subscriptions/api/v1/views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py:45]: done

File: openedx/features/subscriptions/api/v1/views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py]: done

File: openedx/features/subscriptions/models.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/models.py:33]: I think we can rename this field to `max_courses`
```python
max_allowed_courses = models.PositiveIntegerField(
    help_text=_(
        'Maximum number of courses allowed to enroll in paid tracks using the subscription.'
    )
    null=True, 
    blank=True
)
```

File: openedx/features/subscriptions/admin.py
Comment (zubair-arbi) [reply openedx/features/subscriptions/admin.py:15]: Order of fields should be:
`['username', 'subscription_type', 'expiration_date', 'max_allowed_courses', ', 'enrolled_course_ids']`

Also `filter_horizontal` for course enrollments for a better view. (https://stackoverflow.com/questions/5385933/a-better-django-admin-manytomany-field-widget, https://stackoverflow.com/questions/25776235/many-to-many-fields-view-on-django-admin/25776280)

File: openedx/features/subscriptions/api/v1/tests/test_models.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/api/v1/tests/test_models.py:13]: Remove these static variables and use `UserSubscription` constants, like `UserSubscription.LIMITED_ACCESS`

Update everywhere

File: openedx/features/subscriptions/api/v1/tests/test_models.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/api/v1/tests/test_models.py:59]: udpate docstring.

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
I have added feedback for second review.

File: openedx/features/subscriptions/api/v1/views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py]: done

File: openedx/features/subscriptions/admin.py
Comment (rehan99000) [reply openedx/features/subscriptions/admin.py]: done

File: openedx/features/subscriptions/api/v1/serializers.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/serializers.py]: done

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py]: done

File: openedx/features/subscriptions/api/v1/tests/test_models.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/tests/test_models.py:59]: done

File: openedx/features/subscriptions/api/v1/tests/test_models.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/tests/test_models.py:13]: There was an issue with DB access while I was writing tests. I thought it was because of using model constants outside of test functions but turned out to be something else. removed these.

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py:95]: If everything is correctly working, this function should return only one valid subscription even if `filter` is being used. I have used `filter` just so that if something goes wrong, the api doesn't completely crash. If u suggest that I enforce a check on count here, that can be done.

Comment (rehan99000) [reply openedx/features/subscriptions/models.py:102]: Also, unless we write an `APIView`, we have to provide a queryset to be able to serialize it, if we return only on subscription in this, it raises error. I have added a check and log however to ensure that if valid count is one, it logs it and returns empty result.

File: openedx/features/subscriptions/utils.py
Comment (rehan99000) [reply openedx/features/subscriptions/utils.py]: the file wasn't saved. done now

File: openedx/features/subscriptions/api/v1/views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py:48]: done

File: openedx/features/subscriptions/admin.py
Comment (zubair-arbi) [reply openedx/features/subscriptions/admin.py]: Done.

File: lms/envs/common.py
Comment (zubair-arbi) [reply lms/envs/common.py:2336]: As discussed, for now it will be very difficult to create a separate subscription plugin due to subscription code's dependability on core ECOM, LMS code. We will discuss and achieve this requirement in a separate story.

File: openedx/features/subscriptions/api/v1/tests/test_models.py
Comment (ziafazal) [inline openedx/features/subscriptions/api/v1/tests/test_models.py:11]: @rehan99000 
Can we remove pytest related stuff from python test files? Should not it be handled via configs? Did you try running tests without these markers?

File: openedx/features/subscriptions/api/v1/tests/test_views.py
Comment (ziafazal) [inline openedx/features/subscriptions/api/v1/tests/test_views.py:19]: Can we avoid edly related stuff here?

**Review by ziafazal — COMMENTED** [review]
@rehan99000 couple of questions.

File: openedx/features/subscriptions/models.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/models.py:100]: Is this method will return empty list (in case of multiple valid subscriptions) then the learner can keep buys new subscriptions. We should return the latest valid subscription in this case.

File: openedx/features/subscriptions/models.py
Comment (zubair-arbi) [reply openedx/features/subscriptions/models.py:102]: In this case we need to add `Returns` section in the docstring.

File: openedx/features/subscriptions/utils.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/utils.py:16]: `valid_user_subscription` is a queryset so how we can directly use `valid_user_subscription.subscription_id`? shouldn't it be something like `valid_user_subscription.first().subscription_id`?

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
I have added my third review.

File: openedx/features/subscriptions/api/v1/tests/test_views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/tests/test_views.py:19]: @ziafazal To test we need to make API calls. API calls won't succeed otherwise due to our multisite stuff. One solution that comes to mind is to mock the responses.

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py:100]: So should I just log a warning in case a fault occurs? shouldn't there be a way of letting the system/user know something went wrong?

File: openedx/features/subscriptions/api/v1/tests/test_models.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/tests/test_models.py]: removed the marker

File: openedx/features/subscriptions/utils.py
Comment (rehan99000) [reply openedx/features/subscriptions/utils.py:16]: I had modified `get_valid_subscription` method to return just one subscription when I wrote this function. the logic changed afterwards. I have fixed this.

File: openedx/features/subscriptions/api/v1/tests/test_views.py
Comment (zubair-arbi) [reply openedx/features/subscriptions/api/v1/tests/test_views.py:19]: @ziafazal @rehan99000 since we have our edly access middleware enabled default so it will be trivial to use edly namespace here.

@rehan99000 we can try patching some edly access middleware methods in order to avoid `EdlySubOrganizationFactory` usage. Example from ECOM:

```python
@mock.patch('django.contrib.sites.middleware.get_current_site', mock.Mock(return_value=None))
```

File: openedx/features/subscriptions/models.py
Comment (zubair-arbi) [reply openedx/features/subscriptions/models.py]: I think for now logging a warning is fine.

File: openedx/features/subscriptions/api/v1/tests/test_views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/tests/test_views.py:19]: secondly, a much more subtle change would be to mock the edly middleware access method. we can remove this in future while pushing it upstream

Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/tests/test_views.py]: @ziafazal Tests updated. Edly stuff removed

File: openedx/features/subscriptions/api/v1/tests/test_models.py
Comment (ziafazal) [inline openedx/features/subscriptions/api/v1/tests/test_models.py:3]: @rehan99000 remove this `pytest` import too

**Review by ziafazal — APPROVED** [review]
👍

File: openedx/features/subscriptions/api/v1/tests/test_models.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/tests/test_models.py]: done

**Review by iam-mhaseeb — APPROVED** [review]
Looks good to me 👍

File: openedx/features/subscriptions/api/v1/views.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/api/v1/views.py:33]: Add ordering on class level, just like we did for `SubscriptionRetrieveUpdateView` (https://www.django-rest-framework.org/api-guide/filtering/#specifying-which-fields-may-be-ordered-against):
```python
    filter_backends = [filters.OrderingFilter]
    ordering_fields = ['username', 'email']
```

File: openedx/features/subscriptions/api/v1/views.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/api/v1/views.py:40]: Ordering of the fields should be (Please make sure we are following this order for other classes as well, e.g. `SubscriptionsListView`):
```python
class SubscriptionRetrieveUpdateView(PutAsCreateMixin, RetrieveUpdateAPIView):
    """
    Retrieve, update, or create UserSubscriptions.
    """
    authentication_classes = (JwtAuthentication, OAuth2Authentication, SessionAuthentication, )
    permission_classes = (ApiKeyOrModelPermission, )
    model = UserSubscription
    queryset = UserSubscription.objects.all()
    serializer_class = UserSubscriptionSerializer
    lookup_field = 'subscription_id'
    lookup_url_kwarg = 'subscription_id'
    ordering_fields = ['created', ]
    ordering = ['-created']
    pagination_class = None
    page_size = 20
```

File: openedx/features/subscriptions/models.py
Comment (zubair-arbi) [inline openedx/features/subscriptions/models.py:87]: Since this method is now returning a queryset (a list of items) so now we need to update its name `get_valid_subscriptions` and docstring `Get valid subscription of a user.`:
```python
def get_valid_subscriptions(self, user_id=None):        """
        Get valid subscriptions of a user.

        Returns:
            queryset: Returns latest valid user subscription.
        """
```

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
I have added some more feedback.

File: openedx/features/subscriptions/api/v1/views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py:40]: Doesn't make sense to add pagination related stuff in a `RetrieveUpdateAPIView`.

File: openedx/features/subscriptions/api/v1/views.py
Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py:33]: Although this is a List view, it would ideally only return 0 or 1 subscriptions for a given user, doesn't make sense to order it in any way.

File: openedx/features/subscriptions/models.py
Comment (rehan99000) [reply openedx/features/subscriptions/models.py:87]: done

File: openedx/features/subscriptions/api/v1/views.py
Comment (zubair-arbi) [reply openedx/features/subscriptions/api/v1/views.py:40]: Its fine. I just wanted to show you a proper order of Class fields.

File: openedx/features/subscriptions/api/v1/views.py
Comment (zubair-arbi) [reply openedx/features/subscriptions/api/v1/views.py]: Since this is a list view so I would still prefer the explicit ordering.

Comment (rehan99000) [reply openedx/features/subscriptions/api/v1/views.py]: done. I don't think we need the `OrderingFilter`. I have done it like it was done in the retrieve update view

**Review by zubair-arbi — APPROVED** [review]
Thanks for addressing all the feedback. Changes look good now. 👍

**Pull request merged**
